### ISSUE TITLE
[UBSAN] Undefined behavior in DQM for diff relvals

### ISSUE THREAD

**Issue by mrodozov** [cms-sw/cmssw#35004]
In few DQM packages undefined behaviour is reported by the UBSAN IB:

DQM/CSCMonitorModule/plugins/CSCDQM_StripClusterFinder.cc:268:77: - runtime error: member access within address 0x2b14cd588b00 which does not point to an object of type 'StripClusterFitData'
136.885

DQM/HcalCommon/src/DQClient.cc:109:13 - runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'
11646.0
11723.17

DQM/L1TMonitor/src/L1TGMT.cc:205:40: runtime error: signed integer overflow: 117903231 * 3564 cannot be represented in type 'int'
134.99603

DQM/L1TMonitor/src/L1TStage2EMTF.cc:979:40: runtime error: index 6 out of bounds for type 'MonitorElement *[6]'
136.804
136.873

DQMServices/Core/interface/DQMOneEDAnalyzer.h

lines:
31, 38 39 46 48 49 55 58 63
runtime error: member call on address 0x2adec9dd57e0 which does not point to an object of type 'EDProducerBase'

13.0_QCD_Pt_3000_3500 s3
11603.0_SingleElectronPt1000+202 step3

Reference is the latest UBSAN IB:
https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/relVal/CMSSW_12_1/2021-08-20-2300?selectedArchs=slc7_amd64_gcc10&selectedFlavors=UBSAN_X&selectedStatus=failed

**Comment by mrodozov** [comment]
assign dqm

**Comment by makortel** [comment]
> DQM/HcalCommon/src/DQClient.cc:109:13 - runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'

This points to line
https://github.com/cms-sw/cmssw/blob/3a5e558609b0e6f30f9001479d3038ec773a016e/DQM/HcalCommon/src/DQClient.cc#L109
and `_totalLS` being uninitialized.

> DQM/L1TMonitor/src/L1TStage2EMTF.cc:979:40: runtime error: index 6 out of bounds for type 'MonitorElement *[6]'

https://github.com/cms-sw/cmssw/blob/3a5e558609b0e6f30f9001479d3038ec773a016e/DQM/L1TMonitor/src/L1TStage2EMTF.cc#L978-L980
hinting something could be wrong in the `iRPC` calculation

> `DQMServices/Core/interface/DQMOneEDAnalyzer.h `
> lines:
> 31, 38 39 46 48 49 55 58 63
> `runtime error: member call on address 0x2adec9dd57e0 which does not point to an object of type 'EDProducerBase'`

These are all lines with a call to a member function of the `EDProducerBase` base class via explicit `this` pointer, like here
https://github.com/cms-sw/cmssw/blob/3a5e558609b0e6f30f9001479d3038ec773a016e/DQMServices/Core/interface/DQMOneEDAnalyzer.h#L29-L32
It is not immediately clear to me what would be wrong here.

**Comment by makortel** [comment]
> > `DQMServices/Core/interface/DQMOneEDAnalyzer.h `
> > lines:
> > 31, 38 39 46 48 49 55 58 63
> > `runtime error: member call on address 0x2adec9dd57e0 which does not point to an object of type 'EDProducerBase'`
> 
> These are all lines with a call to a member function of the `EDProducerBase` base class via explicit `this` pointer, like here
> https://github.com/cms-sw/cmssw/blob/3a5e558609b0e6f30f9001479d3038ec773a016e/DQMServices/Core/interface/DQMOneEDAnalyzer.h#L29-L32
> 
> 
> It is not immediately clear to me what would be wrong here.

Here is some discussion https://stackoverflow.com/questions/57294792/c-ubsan-produces-false-positives-with-derived-objects with a suggestion

> TLDR: Make sure all classes containing virtual functions are exported when compiling with `-fvisibility=hidden`

**Comment by mrodozov** [comment]
>  and `_totalLS` being uninitialized.

I'll check how many of those will (hopefully) be solved with initialization after I'm done with this issues.
another example of this is here:
https://github.com/dpiparo/vdt/blob/master/include/sincos.h#L135-L139
https://github.com/cms-sw/cmssw/issues/35003
`int j` is uninit
although I'm not sure what details::reduce2quadrant does with `j` (it probably sets it to something)

**Comment by abhih1** [comment]
Hello, regarding the error from `Validation/EcalDigis/plugins/EcalSelectiveReadoutValidation.cc` I tried the suggestions by @makortel, i.e changing the order of the struct member `simHit` after `eta` and also commented out the occurrences of the `bool[]` arrays to see if the error went away, but the error still remains on testing with the workflow `11603.0`. Any more suggestions on fixing this would be very helpful.

**Comment by mrodozov** [comment]
well I have one suggestion and it is to narrow the size artificially using the order of magnitudes from 360 to 36 and 170 to 17 although this wont be correct from precision perspective it will drop the size of that objects 100 times by my calculations.
If the problem is still there with object size just 0.4 the mb it's probably not the size, which we can think of how to drop further to be extra sure

**Comment by abhih1** [comment]
It indeed looks like a memory issue. Changing the large arrays to vectors seems to make the error go away, by allocating memory on the heap as @makortel suggested. I have made a PR for this: https://github.com/cms-sw/cmssw/pull/35130
